--- ./src/init.c.build	2010-11-12 04:00:58.000000000 +0000
+++ ./src/init.c	2010-11-12 04:01:16.000000000 +0000
@@ -8,6 +8,7 @@
 #include <stdio_ext.h>
 #include <dlfcn.h>
 #include <sys/vfs.h>
+#include <sys/stat.h>
 #include <stdint.h>
 #include <limits.h>
 
--- ./src/label_file.c.build	2010-11-12 04:01:29.000000000 +0000
+++ ./src/label_file.c	2010-11-12 04:02:34.000000000 +0000
@@ -17,6 +17,7 @@
 #include <errno.h>
 #include <limits.h>
 #include <regex.h>
+#include <sys/stat.h>
 #include "callbacks.h"
 #include "label_internal.h"
 
--- ./src/Makefile.build	2009-03-02 18:14:53.000000000 +0000
+++ ./src/Makefile	2010-11-12 04:05:33.000000000 +0000
@@ -73,7 +73,7 @@
 	$(CC) $(filter-out -Werror,$(CFLAGS)) -I$(RUBYINC) -fPIC -DSHARED -c -o $@ $<
 
 $(SWIGSO): $(SWIGLOBJ)
-	$(CC) $(CFLAGS) $(LDFLAGS) -shared -o $@ $< -L. -lselinux -L$(LIBDIR) -Wl,-soname,$@
+	$(CC) $(CFLAGS) $(LDFLAGS) -shared -o $@ $< -L. -lselinux `python-config --libs` -L$(LIBDIR) -Wl,-soname,$@
 
 $(SWIGRUBYSO): $(SWIGRUBYLOBJ)
 	$(CC) $(CFLAGS) $(LDFLAGS) -shared -o $@ $^ -L. -lselinux -L$(LIBDIR) -Wl,-soname,$@
@@ -86,7 +86,7 @@
 	$(CC) $(CFLAGS) -I$(PYINC) -fPIC -DSHARED -c -o $@ $<
 
 $(AUDIT2WHYSO): audit2why.lo
-	$(CC) $(CFLAGS) $(LDFLAGS) -shared -o $@ $^ -L. -lselinux ${LIBDIR}/libsepol.a -L$(LIBDIR) -Wl,-soname,$@
+	$(CC) $(CFLAGS) $(LDFLAGS) -shared -o $@ $^ -L. -lselinux `python-config --libs` ${LIBDIR}/libsepol.a -L$(LIBDIR) -Wl,-soname,$@
 
 %.o:  %.c policy.h
 	$(CC) $(CFLAGS) $(TLSFLAGS) -c -o $@ $<
